File: frontend/src/types/progressItem.ts
export type ProgressItem = {
  id: number
  project_id: number
  batch_id: number
  task_id?: number | null
  baseline_plan_id?: number | null
  wbs_code?: string | null
  task_code?: string | null
  task_name?: string | null
  area?: string | null
  construction_unit?: string | null
  building?: string | null
  floor?: string | null
  discipline?: string | null
  system_name?: string | null
  unit?: string | null
  total_quantity?: number | null
  planned_quantity?: number | null
  period_quantity?: number | null
  cumulative_quantity?: number | null
  actual_quantity?: number | null
  remaining_quantity?: number | null
  planned_percent?: number | null
  imported_planned_percent?: number | null
  actual_percent?: number | null
  reported_percent?: number | null
  time_planned_percent?: number | null
  progress_deviation?: number | null
  schedule_phase?: string | null
  current_period_quantity?: number | null
  current_period_percent?: number | null
  planned_start_date?: string | null
  planned_finish_date?: string | null
  actual_start_date?: string | null
  actual_finish_date?: string | null
  weight?: number | null
  value_amount?: number | null
  status?: string | null
  remark?: string | null
  extra_fields?: string | null
  is_manually_edited: boolean
  manual_edit_reason?: string | null
  created_at: string
  updated_at: string
}

export type ProgressItemListResponse = {
  items: ProgressItem[]
  total: number
  page: number
  page_size: number
  scope_info?: ProgressItemScopeInfo | null
}

export type ProgressItemScopeInfo = {
  scope: 'project' | 'batch' | string
  data_date?: string | null
  import_group_id?: string | null
  included_batch_ids: number[]
  included_sheets: string[]
  task_count: number
  message?: string | null
}

export type ProgressItemPayload = {
  reason: string
  actual_quantity?: number | null
  cumulative_quantity?: number | null
  period_quantity?: number | null
  planned_quantity?: number | null
  total_quantity?: number | null
  actual_percent?: number | null
  planned_percent?: number | null
  reported_percent?: number | null
  remaining_quantity?: number | null
  planned_start_date?: string | null
  planned_finish_date?: string | null
  actual_start_date?: string | null
  actual_finish_date?: string | null
  weight?: number | null
  value_amount?: number | null
  status?: string | null
  remark?: string | null
}

export type ProgressItemEditHistory = {
  id: number
  progress_item_id: number
  field_name: string
  old_value?: string | null
  new_value?: string | null
  reason?: string | null
  edited_by?: string | null
  edited_at: string
}

export type ProgressItemFilterOptions = {
  construction_units: string[]
  buildings: string[]
  floors: string[]
  disciplines: string[]
  system_names: string[]
  statuses: string[]
  floors_by_building: Record<string, string[]>
}

File: backend/app/schemas/progress_item.py
from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, Field


class ProgressItemRead(BaseModel):
    id: int
    project_id: int
    batch_id: int
    task_id: int | None = None
    baseline_plan_id: int | None = None
    wbs_code: str | None = None
    task_code: str | None = None
    task_name: str | None = None
    area: str | None = None
    construction_unit: str | None = None
    building: str | None = None
    floor: str | None = None
    discipline: str | None = None
    system_name: str | None = None
    unit: str | None = None
    total_quantity: float | None = None
    planned_quantity: float | None = None
    period_quantity: float | None = None
    cumulative_quantity: float | None = None
    actual_quantity: float | None = None
    remaining_quantity: float | None = None
    planned_percent: float | None = None
    imported_planned_percent: float | None = None
    actual_percent: float | None = None
    reported_percent: float | None = None
    time_planned_percent: float | None = None
    progress_deviation: float | None = None
    schedule_phase: str | None = None
    current_period_quantity: float | None = None
    current_period_percent: float | None = None
    planned_start_date: date | None = None
    planned_finish_date: date | None = None
    actual_start_date: date | None = None
    actual_finish_date: date | None = None
    weight: float | None = None
    value_amount: float | None = None
    status: str | None = None
    remark: str | None = None
    extra_fields: str | None = None
    is_manually_edited: bool
    manual_edit_reason: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ProgressItemScopeInfo(BaseModel):
    scope: str
    data_date: date | None = None
    import_group_id: str | None = None
    included_batch_ids: list[int] = []
    included_sheets: list[str] = []
    task_count: int = 0
    message: str | None = None


class ProgressItemListResponse(BaseModel):
    items: list[ProgressItemRead]
    total: int
    page: int
    page_size: int
    scope_info: ProgressItemScopeInfo | None = None


class ProgressItemUpdate(BaseModel):
    reason: str = Field(..., min_length=1)
    actual_quantity: float | None = None
    cumulative_quantity: float | None = None
    period_quantity: float | None = None
    planned_quantity: float | None = None
    total_quantity: float | None = None
    actual_percent: float | None = None
    planned_percent: float | None = None
    reported_percent: float | None = None
    remaining_quantity: float | None = None
    planned_start_date: date | None = None
    planned_finish_date: date | None = None
    actual_start_date: date | None = None
    actual_finish_date: date | None = None
    weight: float | None = None
    value_amount: float | None = None
    status: str | None = None
    remark: str | None = None


class ProgressItemFilterOptions(BaseModel):
    construction_units: list[str] = []
    buildings: list[str] = []
    floors: list[str] = []
    disciplines: list[str] = []
    system_names: list[str] = []
    statuses: list[str] = []
    floors_by_building: dict[str, list[str]] = {}


class ProgressItemEditHistoryRead(BaseModel):
    id: int
    progress_item_id: int
    field_name: str
    old_value: str | None = None
    new_value: str | None = None
    reason: str | None = None
    edited_by: str | None = None
    edited_at: datetime

    model_config = {"from_attributes": True}
